### ISSUE TITLE
Non-functional binaries in PATH containing spaces

### ISSUE THREAD

**Issue by mojca** [asciidoctor/asciidoctor-diagram#266]
I happened to let `gnuplot` do the default installation under `C:\Program Files\gnuplot\bin\gnuplot.exe`.

However when running asciidoctor on the file it seems to get stuck at https://github.com/asciidoctor/asciidoctor-diagram/blob/5e7dcfb7e9251502b6499ccc8410de5b651e9dde/lib/asciidoctor-diagram/util/cli.rb#L64 due to `*args` (`["C:\\Program Files\\gnuplot\\bin/gnuplot.EXE"]'`) being split at whitespace:
```
> asciidoctor -r asciidoctor-diagram test.adoc
asciidoctor: ERROR: test.adoc: line 10: Failed to generate image: No such file or directory - C:\Program Files\gnuplot\bin/gnuplot.EXE
```

**Comment by mojavelinux** [comment]
I believe if the command is passed as an array instead of a string, it will not be parsed. That's the change that should be applied to fix this problem.

**Comment by mojavelinux** [comment]
In fact, it works if there's at least one empty argument:

```
Open3.capture3 './call me.sh', ''
```

or the shell is disabled:

```
Open3.capture3 ['./call me.sh', '']
```

If there are no arguments, then it tries to parse the first parameter.

Another way to do it is to use Shellwords:

```
Open3.capture3 Shellwords.shellescape './call me.sh'
```

**Comment by mojca** [comment]
I'm not familiar with ruby, but I would tend to almost call this a bug in ruby.

The documentation of `capture3` never explicitly promises to handle arrays, but it says it forwards parameters to `popen3` and `popen3` docs kind of says that it should work:
```ruby
Open3.popen3(["echo", "argv0"], "abc") {|i, o, e, t| ... }
```
even though reading the source code doesn't make one any more optimistic.

**Comment by mojavelinux** [comment]
> I would tend to almost call this a bug in ruby.

It's certainly a very surprising "feature". We'll need to watch out for this.

**Comment by pepijnve** [comment]
The error message is `No such file or directory - C:\Program Files\gnuplot\bin/gnuplot.EXE`. So `capture3` is not splitting the argument; it just can't resolve the given path. Most likely this is due to the mix of `/` and `\`. I'm not sure which one `capture3` expects. I'll need to find some time to revive my Windows VM to figure this one out.

**Comment by mojca** [comment]
I don't think you need a Windows VM to reproduce the problem.
```bash
$ find .
.
./test.rb
./test1 dir
./test1 dir/test.sh

$ cat test.rb
#!/usr/bin/env ruby
require "open3"
o, e, s = Open3.capture3('test1 dir/test.sh', :stdin_data=>"Hello")

$ ruby test.rb
Traceback (most recent call last):
        4: from test.rb:3:in `<main>'
        3: from /usr/lib/ruby/2.5.0/open3.rb:258:in `capture3'
        2: from /usr/lib/ruby/2.5.0/open3.rb:95:in `popen3'
        1: from /usr/lib/ruby/2.5.0/open3.rb:199:in `popen_run'
/usr/lib/ruby/2.5.0/open3.rb:199:in `spawn': No such file or directory - test1 (Errno::ENOENT)
```

**Comment by mojca** [comment]
Also note that when `gnuplot` is located in some path without spaces, asciidoctor-diagram would pass `C:\\Programs\\gnuplot\\bin/gnuplot.EXE` (as printed from `puts`, so backslashes are just escaped) to `capture3` and that one works flawlessly. So I'm pretty sure the issue is not with (back)slashes.

**Comment by pepijnve** [comment]
Observe the difference in errors:
`No such file or directory - test1`
versus
`No such file or directory - C:\Program Files\gnuplot\bin/gnuplot.EXE`

In the first case the string has been split and the error only shows the first part of the string. In the second case the entire path is shown, including the space between Program and Files. The split on the space did not happen. The space split thing is a red herring; the code is already passing the arguments correctly to get the desired effect.

Let me run this through the debugger and see what's going on.

**Comment by pepijnve** [comment]
So the issue was and wasn't the space 😄 

The major gotcha you and @mojavelinux hinted at is the difference between the three variants of Kernel#spawn
```
1. commandline                 : command line string which is passed to the standard shell
2. cmdname, arg1, ...          : command name and one or more arguments (This form does not use the shell. See below for caveats.)
3. [cmdname, argv0], arg1, ... : command name, argv[0] and zero or more arguments (no shell)
```

Most of the commands were using variant 2 since they pass command line arguments. For Gnuplot there are no arguments for the command and as a consequence we were triggering variant 1 instead. One uses shell based execution, the other does not 🤦‍♂

The fix in bf7e169 forces variant 3 in all situations to avoid this pitfall.
